### DISCUSSION TITLE
Computing millions of intersections at once causes high residual memory usage?

### DISCUSSION THREAD

**Discussion by SnoopJ** [shapely/shapely#1828] [Q&A] [answered]
I'm using `shapely` (v2.0.1 on Python 3.7.12) to compute millions of intersections at a time by broadcasting two arrays of boxes against each other, and I've found that if I do this as one large computation, the resident set of the Python process after the computation remains high, but if I break my computation into batches, I can keep the size of the resident set small. See reproduction and corresponding output below:

<details><summary>Click for reproduction program</summary>

```python
import gc
import os
import time

import numpy as np
import shapely


PID = os.getpid()
np.random.seed(42)

def random_boxes(N: int):
    boxW = 10 + 20*np.random.random_sample(size=N)
    boxH = 10 + 20*np.random.random_sample(size=N)

    xmin = 128*np.random.random_sample(size=N)
    ymin = 128*np.random.random_sample(size=N)
    xmax = xmin + boxW
    ymax = ymin + boxH

    return shapely.box(xmin, ymin, xmax, ymax)


def memory_usage_kb() -> int:
    with open(f"/proc/{PID}/status", "r") as f:
        mem_line = next(line for line in f if line.startswith("VmRSS:"))
        mem_line = mem_line[len("VmRSS:"):-3]
        mem_kB = int(mem_line)
    return mem_kB


# additional axes here are so that broadcasting gives us every one of the 
# 2000*2500 = 5M combinations of these boxes
boxes_A = random_boxes(2000)[:, np.newaxis]
boxes_B = random_boxes(2500)[np.newaxis, :]


# number of boxes from A to consider at once
for BATCH_A_SIZE in (100, 2000):
    N_BATCHES = boxes_A.size // BATCH_A_SIZE
    if N_BATCHES == 0 or (boxes_A.size > BATCH_A_SIZE and boxes_A.size % BATCH_A_SIZE != 0):
        N_BATCHES += 1

    print(f"Computing intersections with BATCH_A_SIZE={BATCH_A_SIZE}, N_BATCHES={N_BATCHES}")

    tottime = 0
    for batchnum in range(N_BATCHES):
        a_idx = slice(batchnum*BATCH_A_SIZE, (batchnum+1)*BATCH_A_SIZE)
        Nboxes = np.broadcast(boxes_A[a_idx, :], boxes_B).size
        print(f"Computing {Nboxes} intersections in batch #{batchnum+1}/{N_BATCHES}…")
        shapely.intersection(boxes_A[a_idx, :], boxes_B)

    # sleep for a few seconds and force GC just to be sure we aren't seeing
    # memory usage from dead objects
    time.sleep(3)
    gc.collect()

    print(f"Current VmRSS: {memory_usage_kb()/1000} MB")
```

</details> 

```
$ venv/bin/python3 repro.py 
Computing intersections with BATCH_A_SIZE=100, N_BATCHES=20
Computing 250000 intersections in batch #1/20…
/tmp/shapely_sandbox/venv/lib/python3.7/site-packages/shapely/set_operations.py:133: RuntimeWarning: invalid value encountered in intersection
  return lib.intersection(a, b, **kwargs)
Computing 250000 intersections in batch #2/20…
Computing 250000 intersections in batch #3/20…
Computing 250000 intersections in batch #4/20…
Computing 250000 intersections in batch #5/20…
Computing 250000 intersections in batch #6/20…
Computing 250000 intersections in batch #7/20…
Computing 250000 intersections in batch #8/20…
Computing 250000 intersections in batch #9/20…
Computing 250000 intersections in batch #10/20…
Computing 250000 intersections in batch #11/20…
Computing 250000 intersections in batch #12/20…
Computing 250000 intersections in batch #13/20…
Computing 250000 intersections in batch #14/20…
Computing 250000 intersections in batch #15/20…
Computing 250000 intersections in batch #16/20…
Computing 250000 intersections in batch #17/20…
Computing 250000 intersections in batch #18/20…
Computing 250000 intersections in batch #19/20…
Computing 250000 intersections in batch #20/20…
Current VmRSS: 90.72 MB
Computing intersections with BATCH_A_SIZE=2000, N_BATCHES=1
Computing 5000000 intersections in batch #1/1…
Current VmRSS: 1057.748 MB
```

This high residual memory usage seems to persist, even if I delete any retained references to array objects/etc. in my program, forcibly remove `shapely` from `sys.modules`, etc.

I plan to run this code in a process that is already pretty memory-hungry, so the prospect of losing O(1 GB) of headroom is kind of concerning. Profiling this program with `memray` shows me that most of the allocations happen in [`ufuncs.c:YY_Y_func()`](https://github.com/shapely/shapely/blob/5c4670bbbac54fffddbabac202efdecdc445317d/src/ufuncs.c#L910), but as far as I can tell, this function is not relying on `pymalloc` and seems to clean up after itself.

Anybody know what causes the high residual with the all-at-once approach? Is this maybe a case of CPython's "high water mark" memory behavior? The batched approach works for me, but I'm wondering why the memory usage differs in these two cases.

**Comment by jorisvandenbossche** [comment]
Thanks for the nice reproducible example!
I can confirm that I see the same behaviour, but I am no expert on those low-level memory details to really understand or explain this.

Using memray, we can see that most of the memory allocations are done by GEOS, as expected, by `GEOSIntersection_r` (which is called inside `YY_Y_func`). This is the only place where so much memory is being allocated, so if there is residual memory usage, I would expect it to come from here. 

The graph from memray indicates that the heap allocated memory goes down at the end:

![newplot](https://github.com/shapely/shapely/assets/1020496/d8820f98-2ecf-42a8-bfdf-64b2b7d4c96a)

One thing to note is that _most_ of the intersections are actually empty polgyons. But those seem to still take a sizable portion of the memory allocations from GEOSIntersection (700MB from `geos::operation::overlayng::OverlayNG::createEmptyResult()`). This seems a bit strange.

**Reply by jorisvandenbossche** [reply]
As another test, if we repeat the for loop in your script a few times, memray shows the following timeline:

![newplot(1)](https://github.com/shapely/shapely/assets/1020496/0f3b0913-2643-4874-892f-e831c54d3e3b)

So it's not that the memory usage keeps increasing every loop. So I assume that the high resident memory usage at the end is mostly because the system doesn't yet show it as freed? (but not being an actual problem or memory leak)

> One thing to note is that _most_ of the intersections are actually empty polgyons. But those seem to still take a sizable portion of the memory allocations 

Nonetheless, if you would like to reduce the actual memory usage of this program (and probably improve the speed as well), you could use a STRtree to only calculate the intersections of geometry combinations that would actually result in a non-empty geometry, instead of calculating the intersection for the full product-wise combinations of all geometries (less than 10% of the combinations result in a non-empty result, so that way you can avoid the creation of many empty polygons, which because of the memory model of GEOS still occupy quite some memory. I don't know if this is similar in your real use case, or only specific to the reproducer, of course).  

```python
# don't reshape them to 2D arrays (which was done to use broadcasting for all combinations) 
boxes_A = random_boxes(2000)
boxes_B = random_boxes(2500)

tree = shapely.STRtree(boxes_B)
idxs = tree.query(boxes_A, predicate="intersects")
shapely.intersection(boxes_A[idxs[0]], boxes_B[idxs[1]])
```

Timing both:

```
In [5]: %%time
   ...: res1 = shapely.intersection(boxes_A[:, np.newaxis], boxes_B[np.newaxis, :])
CPU times: user 17.2 s, sys: 136 ms, total: 17.4 s
Wall time: 17.4 s

In [7]: %%time
   ...: tree = shapely.STRtree(boxes_B)
   ...: idxs = tree.query(boxes_A, predicate="intersects")
   ...: res2 = shapely.intersection(boxes_A[idxs[0]], boxes_B[idxs[1]])
CPU times: user 6.11 s, sys: 104 ms, total: 6.21 s
Wall time: 6.25 s

**Reply by SnoopJ** [reply]
> avoid the creation of many empty polygons, which because of the memory model of GEOS still occupy quite some memory. I don't know if this is similar in your real use case, or only specific to the reproducer, of course

You've intuited part of the real problem, which is computing intersections of known truth and predicted boxes in an object detection model! I actually discovered this issue after writing a STRtree solution for Shapely 1.8 was not enough speedup. I hadn't considered that STRtree supports the same array-like input though, so my solution will definitely include that. I am a little worried about the copying the "advanced indexing" can cause there, but it may be comparable to the copies the broadcasting solution causes.

The heap size going down without the resident size decreasing is kind of fascinating, I'm not sure how this is even possible! Possibly memray is reporting on some subset of the heap, or maybe the OS hasn't reclaimed the allocated pages as you suggest. I think I can test the latter theory by running some other process that allocates a ton of memory and see if thay shrinks the RSS of the reproduction program. If that pans out, it may mean that I'm reading too much into the RSS value, which would be a tidy resolution of my confusion about your plots.

Interesting aside: it seems memray is much slower to instrument this program running under 3.7 than it is under 3.11. With the latter I see a runtime of O(10x) uninstrumented, but with the former it's more like O(600x) (both using the --native flag which is a bit more demanding AIUI). I'm not positive if the same RSS/heap profile shows up in that case, but I plan to have a look soon.

Thank you for your replies, it helps a lot to have some maintainer perspective here, especially knowing that I'm trying the right things to accelerate my real application. 😁

**Reply by jorisvandenbossche** [reply]
> I actually discovered this issue after writing a STRtree solution for Shapely 1.8 was not enough speedup. I hadn't considered that STRtree supports the same array-like input though, so my solution will definitely include that. 

Yes, the STRtree in 2.0 is much improved compared to shapely 1.8 (support for "bulk" queries with arrays, which is much faster). 

> I am a little worried about the copying the "advanced indexing" can cause there, but it may be comparable to the copies the broadcasting solution causes.

Yes, that will be comparable: it actually does copy (while broadcasting is done with a view), but because the numpy array is an object dtype array with essentially pointers to the Geometry objects, that doesn't cause an actual copy of the geometries itself, only of those pointers. So that gives only negligible overhead (plus that it ensures you need to calculate less intersections, which is more costly).

**Reply by SnoopJ** [reply]
> I think I can test the latter theory by running some other process that allocates a ton of memory and see if thay shrinks the RSS of the reproduction program.

For the sake of completeness, this did result in some very small reclamation of memory used by the process, but the reclaimed memory only amounted to a handful of 4 KiB pages, a drop in the bucket. Getting this negative result helped lead me to [the real answer](https://github.com/shapely/shapely/discussions/1828#discussioncomment-6091427)

**Comment by SnoopJ** [comment, marked as answer]
I think I've worked out where the residual RSS usage comes from: `glibc`'s allocator!

### **tl;dr** - so many small allocations badly fragments the heap, calling `malloc_trim()` can recover a lot of memory!

I hadn't before realized that the library has [its own arena allocation](https://sourceware.org/glibc/wiki/MallocInternals#Arenas_and_Heaps), but once I learned this detail, everything sorta clicked into place, since fragmentation pretty much perfectly explains the behavior of the reproduction program.

Running the reproduction program with `-i` to keep the process (and its maps) alive, we can run `pmap -x <pid>` on it to discover that the residual memory is being consumed by a single large mapping. Running the program with `strace --trace=%memory` lets us also correlate this large map with the syscalls `glibc` uses to grow it.

```
10:03 [jgerity@giskard /tmp/shapely_sandbox]
$ strace --output=strace_mem.txt --trace=%memory venv/bin/python3 -i repro_discussion.py &
[1] 1139740
10:05 [jgerity@giskard /tmp/shapely_sandbox]
$ Computing intersections with BATCH_A_SIZE=100, N_BATCHES=20
Computing 250000 intersections in batch #1/20…
… snip …
Computing 5000000 intersections in batch #1/1…
Current VmRSS: 1057.44 MB

10:09 [jgerity@giskard /tmp/shapely_sandbox]
$ pmap -x 1139752  # note: separate PID because strace is its own process
1139752:   venv/bin/python3 -i repro_discussion.py
Address           Kbytes     RSS   Dirty Mode  Mapping
0000000000400000     136     136       0 r---- python3.7
0000000000422000    1676    1676       0 r-x-- python3.7
00000000005c5000     720     404       0 r---- python3.7
000000000067a000       4       4       4 r---- python3.7
000000000067b000     420     420     348 rw--- python3.7
00000000006e4000     132     132     132 rw---   [ anon ]
0000000001574000 1027176 1027108 1027108 rw---   [ anon ]    <---- this one right here
00007ff676ef2000     256     256     256 rw---   [ anon ]
00007ff676fb2000     512     512     512 rw---   [ anon ]
… snip …
```

The output of `pmap` also confirms that the residual memory is not being held by `libgeos` or `shapely`'s wrappers thereof.

Looking at the output of this `strace` shows a long string of [`brk()`](https://man7.org/linux/man-pages/man2/brk.2.html) that grows this mapping by 132 KiB at a time, which is apparently a magic number for the heap implementation (4 + 128 KiB).

```
10:15 [jgerity@giskard /tmp/shapely_sandbox]
$ grep -B10 "0x4008e000" strace_mem.txt   # 0x1574000 + 1027176*1024 = 0x4008e000
brk(0x3ff86000)                         = 0x3ff86000
brk(0x3ffa7000)                         = 0x3ffa7000
brk(0x3ffc8000)                         = 0x3ffc8000
brk(0x3ffe9000)                         = 0x3ffe9000
brk(0x4000a000)                         = 0x4000a000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff666e0c000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff666dcc000
brk(0x4002b000)                         = 0x4002b000
brk(0x4004c000)                         = 0x4004c000
brk(0x4006d000)                         = 0x4006d000
brk(0x4008e000)                         = 0x4008e000   <---- (0x8e000 - 0x6d000) / 1024 = 132
```

I did try swapping in the alternate heap implementations offered by `tcmalloc, jemalloc` to see what effect this would have, but the fragmentation and residual memory usage was still present. Since `glibc` apparently creates a separate arena for each thread (if I'm understanding their documentation correctly), I did check if wrapping my interseciton loop up in a `threading.Thread` and running it that way would maybe isolate the fragmentation, but it also has no effect on the residual usage. In hindsight, this program has a perfectly pathological allocation pattern to trigger this kind of fragmentation!

~~I can see the silhouette of hypothetical functionality in shapely/GEOS that might account for my program's utilization pattern, by setting aside a separate bit of memory that can be released all at once for this operation (probably avoiding the libc `malloc()` altogether?), but this is so much of an edge case that it's almost certainly not worth the effort it would require on the part of both libraries, especially since batching the computation prevents thrashing the heap without any increase in overhead.~~

Edit: it turns out that calling [`malloc_trim()`](https://man7.org/linux/man-pages/man3/malloc_trim.3.html) from inside of the reproduction program can recover almost all of the residual memory! Here's the additional code for doing that:

```python
import ctypes
import ctypes.util

libc = ctypes.CDLL(ctypes.util.find_library('c'), use_errno=True)

# … rest of the program runs here …

print("Calling malloc_trim()")
libc.malloc_trim(0)  # argument of 0 means a maximally-aggressive trimming of the heap
print(f"Current VmRSS: {memory_usage_kb()/1000} MB")
```

Having made this change, I see a residual memory usage of only ~50 MB, nice!

```
…
Computing 250000 intersections in batch #20/20…
Current VmRSS: 90.504 MB
Computing intersections with BATCH_A_SIZE=2000, N_BATCHES=1
Computing 5000000 intersections in batch #1/1…
Current VmRSS: 1057.536 MB
Calling malloc_trim()
Current VmRSS: 48.944 MB
```

Thanks for the help in unraveling this mystery, and the suggestions provided along the way! :grin:
